doc/dev_guide/c_style_guide_new.rst

```diff
@@ -0,0 +1,1046 @@
+remember: ``indent`` is not a fix for bad programming.
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Chapter 10: Macros, Enums and RTL
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Names of macros defining constants and labels in enums are capitalized.
+
+.. code-block:: c
+
+	#define CONSTANT 0x12345
+
+Enums are preferred when defining several related constants.
+
+CAPITALIZED macro names are appreciated but macros resembling functions
+may be named in lower case.
+
+Generally, inline functions are preferable to macros resembling functions.
+
+Macros with multiple statements should be enclosed in a do - while block:
+
+.. code-block:: c
+
+  #define macrofun(a, b, c)       \
+    do {                          \
+      if (a == 5)                 \
+        do_this(b, c);            \
+    } while (0)
+
+Things to avoid when using macros:
+
+1) macros that affect control flow:
+
+   .. code-block:: c
+
+     #define FOO(x)                  \
+       do {                          \
+         if (blah(x) < 0)            \
+           return -EBUGGERED;        \
+       } while (0)
+
+   is a **very** bad idea.  It looks like a function call but exits the ``calling``
+   function; don't break the internal parsers of those who will read the code.
+
+2) macros that depend on having a local variable with a magic name:
+
+   .. code-block:: c
+
+     #define FOO(val) bar(index, val)
+
+   might look like a good thing, but it's confusing as hell when one reads the
+   code and it's prone to breakage from seemingly innocent changes.
+
+3) macros with arguments that are used as l-values: FOO(x) = y; will
+   bite you if somebody e.g. turns FOO into an inline function.
+
+4) forgetting about precedence: macros defining constants using expressions
+   must enclose the expression in parentheses. Beware of similar issues with
+   macros using parameters.
+
+   .. code-block:: c
+
+     #define CONSTANT 0x4000
+     #define CONSTEXP (CONSTANT | 3)
+
+5) namespace collisions when defining local variables in macros resembling
+   functions:
+
+   .. code-block:: c
+
+     #define FOO(x)            \
+     ({                        \
+       typeof(x) ret;          \
+       ret = calc_ret(x);      \
+       (ret);                  \
+     })
+
+   ret is a common name for a local variable - __foo_ret is less likely
+   to collide with an existing variable.
+
+   The cpp manual deals with macros exhaustively. The gcc internals manual also
+   covers RTL which is used frequently with assembly language in the kernel.
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Chapter 11: Allocating memory
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. admonition:: Tarantool Style
+    :class: FACT
+
+    Prefer the supplied slab (salloc) and pool (palloc) allocators to malloc()/free()
+    for any performance-intensive or large  memory allocations. Repetitive use of
+    malloc()/free() can lead to memory fragmentation and should therefore be avoided.
```

Comment:
We don't use `salloc` nor `palloc`. Also `malloc` is not a big deal when is used in a non-perf critical code. For perf critical code there are various options and hacks, not only slab allocator.

Reply:
It [came](https://www.tarantool.io/en/doc/latest/dev_guide/c_style_guide/#allocating-memory) from the previous version of this guide. Should I remove this statement at all?

Reply:
I would say something like "prefer specialized allocators like region, mempool, smalloc" instead of "salloc/palloc". For perf critical data, which is almost everything with rare exceptions. "salloc/palloc" didn't ever exist in our code, so the previous version of the guide was also wrong.